File: src/SmeAccounting.Web/Models/Address.cs
using System.Text.RegularExpressions;

namespace SmeAccounting.Web.Models;

public partial class Address : IEquatable<Address>
{
    public string Street { get; }
    public string ProvinceId { get; }
    public string DistrictId { get; }
    public string WardId { get; }

    private Address(string street, string provinceId, string districtId, string wardId)
    {
        Street = street;
        ProvinceId = provinceId;
        DistrictId = districtId;
        WardId = wardId;
    }

    public override string ToString() => $"{Street}, {WardId}, {DistrictId}, {ProvinceId}";

    public static Result<Address> Create(string street, string provinceId, string districtId, string wardId)
    {
        if (string.IsNullOrWhiteSpace(street) || street.Length > 500)
            return Result<Address>.Failure("Street address required, max 500 chars");

        if (string.IsNullOrWhiteSpace(provinceId))
            return Result<Address>.Failure("Province required");

        if (string.IsNullOrWhiteSpace(districtId))
            return Result<Address>.Failure("District required");

        if (string.IsNullOrWhiteSpace(wardId))
            return Result<Address>.Failure("Ward required");

        return Result<Address>.Success(new Address(street, provinceId, districtId, wardId));
    }

    public bool Equals(Address? other) =>
        other is not null && Street == other.Street && ProvinceId == other.ProvinceId
            && DistrictId == other.DistrictId && WardId == other.WardId;

    public override bool Equals(object? obj) => obj is Address other && Equals(other);
    public override int GetHashCode() => HashCode.Combine(Street, ProvinceId, DistrictId, WardId);
}

File: server/src/domain/valueObjects/Address.ts
export interface AddressParts {
  street: string;
  ward?: string;
  district?: string;
  province: string;
  wardCode?: string;
  districtCode?: string;
  provinceCode?: string;
}

export class Address {
  constructor(private readonly parts: AddressParts) {}

  static create(parts: AddressParts): Address {
    if (!parts.street.trim()) throw new Error('Street address is required');
    if (!parts.province.trim()) throw new Error('Province is required');
    return new Address({ ...parts });
  }

  toString(): string {
    const parts = [this.parts.street];
    if (this.parts.ward) parts.push(this.parts.ward);
    if (this.parts.district) parts.push(this.parts.district);
    parts.push(this.parts.province);
    return parts.join(', ');
  }

  get street(): string { return this.parts.street; }
  get ward(): string | undefined { return this.parts.ward; }
  get district(): string | undefined { return this.parts.district; }
  get province(): string { return this.parts.province; }
  get wardCode(): string | undefined { return this.parts.wardCode; }
  get districtCode(): string | undefined { return this.parts.districtCode; }
  get provinceCode(): string | undefined { return this.parts.provinceCode; }

  equals(other: Address): boolean {
    return this.toString() === other.toString();
  }

  toParts(): AddressParts {
    return { ...this.parts };
  }
}
